### PR TITLE
feat!(types,rpc): Add share, row, merkle proofs and share.GetRange

### PR DESCRIPTION
**Pull request by zvolin**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -18,8 +18,8 @@ celestia-tendermint-proto = "0.32.1"
 # Uncomment to apply local changes
 #beetswap = { path = "../beetswap" }
 #blockstore = { path = "../blockstore" }
-#celestia-tendermint = { path = "../celestia-tendermint-rs/tendermint" }
```
Comment (zvolin) [inline Cargo.toml:21]: this needs to be removed when https://github.com/eigerco/celestia-tendermint-rs/pull/31 is merged

File: types/src/merkle_proof.rs [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+                first_leaf_index + subtrees_split,
+                leaf_to_prove,
+                &leaves[subtrees_split..],
+                aunts,
+            );
+
+            // if current subtree has the leaf to prove
+            if (first_leaf_index..first_leaf_index + total).contains(&leaf_to_prove) {
+                if leaf_to_prove < first_leaf_index + subtrees_split {
+                    // leaf was in left subtree, so it's aunt is right hash
```
Comment (fl0rek) [inline types/src/merkle_proof.rs:120]:
```suggestion
                    // leaf was in left subtree, so its aunt is right hash
```

File: types/src/share/proof.rs [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+    }
+}
+
+impl Protobuf<RawShareProof> for ShareProof {}
+
+impl TryFrom<RawShareProof> for ShareProof {
+    type Error = Error;
+
+    fn try_from(value: RawShareProof) -> Result<Self> {
+        println!("{:?}", value.namespace_id);
```
Comment (fl0rek) [inline types/src/share/proof.rs:98]: leftover println

File: types/src/share/proof.rs [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+                &value.namespace_id,
+            )?,
+            share_proofs: value
+                .share_proofs
+                .into_iter()
+                .map(TryInto::try_into)
+                .collect::<Result<_>>()?,
+            row_proof: value
+                .row_proof
+                .ok_or_else(|| validation_error!("todo"))?
```
Comment (fl0rek) [inline types/src/share/proof.rs:120]: leftover todo

File: types/src/share/proof.rs [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+
+    fn try_from(value: RawShareProof) -> Result<Self> {
+        println!("{:?}", value.namespace_id);
+        Ok(Self {
+            data: value
+                .data
+                .into_iter()
+                .map(TryInto::try_into)
+                .collect::<Result<_, _>>()
+                .map_err(|_| validation_error!("todo"))?,
```
[the remark this answers is not in the document]
Comment (zvolin) [reply types/src/share/proof.rs:105]: oopsie

**Review by oblique — APPROVED** [review]
LGTM

**Review by fl0rek — APPROVED** [review]
nice 🎸

**Pull request merged**